aries_cloudagent/resolver/default/peer3.py

```diff
@@ -0,0 +1,193 @@
+        try:
+            async with profile.session() as session:
+                storage = session.inject(BaseStorage)
+                record = await storage.find_record(
+                    BaseConnectionManager.RECORD_TYPE_DID_DOCUMENT, {"did": dp3_doc.id}
+                )
+        except StorageNotFoundError:
+            record = StorageRecord(
+                BaseConnectionManager.RECORD_TYPE_DID_DOCUMENT,
+                dp3_doc.to_json(),
+                {"did": dp3_doc.id},
+            )
+            async with profile.session() as session:
+                storage: BaseStorage = session.inject(BaseStorage)
+                await storage.add_record(record)
+        else:
+            async with profile.session() as session:
+                storage: BaseStorage = session.inject(BaseStorage)
+                await storage.update_record(
+                    record, dp3_doc.to_json(), {"did": dp3_doc.id}
+                )
+        await _reset_keys_from_did_doc(profile, dp3_doc)
+
+        return dp3_doc
+
+
+async def _reset_keys_from_did_doc(profile, did_doc):
+    async with profile.session() as session:
+        storage: BaseStorage = session.inject(BaseStorage)
+        await storage.delete_all_records(
+            BaseConnectionManager.RECORD_TYPE_DID_KEY, {"did": did_doc.id}
+        )
+
+    for vm in did_doc.verification_method or []:
+        if vm.controller == did_doc.id:
+            if vm.public_key_base58:
+                await _add_key_for_did(profile, did_doc.id, vm.public_key_base58)
+            if vm.public_key_multibase:
+                pk = multibase.decode(vm.public_key_multibase)
+                if len(pk) == 32:  # No multicodec prefix
+                    pk = bytes_to_b58(pk)
+                else:
+                    codec, key = multicodec.unwrap(pk)
+                    if codec == multicodec.multicodec("ed25519-pub"):
+                        pk = bytes_to_b58(key)
+                    else:
+                        continue
+                await _add_key_for_did(profile, did_doc.id, pk)
+
+
+async def _add_key_for_did(profile, did: str, key: str):
+    """Store a verkey for lookup against a DID.
+
+    Args:
+        did: The DID to associate with this key
+        key: The verkey to be added
+    """
+    record = StorageRecord(
+        BaseConnectionManager.RECORD_TYPE_DID_KEY, key, {"did": did, "key": key}
+    )
+    async with profile.session() as session:
+        storage: BaseStorage = session.inject(BaseStorage)
+        try:
+            await storage.find_record(
+                BaseConnectionManager.RECORD_TYPE_DID_KEY, {"key": key}
+            )
+        except StorageNotFoundError:
+            await storage.add_record(record)
+        except StorageDuplicateError:
+            pass
+            # "Key already associated with DID: %s; this is likely caused by "
+            # "routing keys being erroneously stored in the past",
```

Comment:
This method is a copy paste from the base connection manager. While I recommended not leaving doc storage and retrieval to the base connection manager, I would recommend leaving the key -> did -> connection management to the base connection manager.

Reply:
Is it appropriate to instantiate an instance of the base manage to call these methods. I wasn't sure if that would look weirder. easy enough change to make if that is the preferred separation of concerns